### PR TITLE
AbstractSSAJumpAggregator interface

### PR DESCRIPTION
**Pull request by alanderos91**
I think everything we discussed is in place. Please feel free to suggest alternative names for the various functions and fields added.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
@isaacsas wanna take a look?

File: src/problem.jl [hunk 1]
```diff
@@ -19,15 +19,16 @@ JumpProblem(prob,aggregator::AbstractAggregatorAlgorithm,jumps::AbstractJump...;
 JumpProblem(prob,jumps::JumpSet;kwargs...) = JumpProblem(prob,NullAggregator(),jumps;kwargs...)
 
 function JumpProblem(prob,aggregator::AbstractAggregatorAlgorithm,jumps::JumpSet;
-                     save_positions = typeof(prob) <: AbstractDiscreteProblem ? (false,true) : (true,true))
+                     save_positions = typeof(prob) <: AbstractDiscreteProblem ? (false,true) : (true,true),
+                     rng = Xorshifts.Xoroshiro128Star(rand(UInt64)))
```
Comment (ChrisRackauckas) [inline src/problem.jl]: Let's use Xoroshiro128Plus. See the end of the chat which states it's not the same as the one in the post.

**Comment by ChrisRackauckas** [comment]
Forgot to define the abstract type

File: src/problem.jl (hunk 1 above)
Comment (isaacsas) [reply src/problem.jl]: Just to clarify. The blog post found issues with Xoroshiro128Plus. It recommended as fine XorShift* 128/64. I haven't found any info about Xoroshiro128star.

Comment (ChrisRackauckas) [reply src/problem.jl]: Oh really? But then it passed all of the tests of the repo?

File: src/aggregators/direct.jl [hunk 2]
```diff
@@ -1,68 +1,94 @@
...
 
-DiscreteCallback(c::DirectJumpAggregation) = DiscreteCallback(c,c,initialize=c,save_positions=c.save_positions)
+######################## SSA specific helper routines ########################
+
+@fastmath function time_to_next_jump(p::DirectJumpAggregation, u, params, t)
+  cur_rates = p.cur_rates
+  rates = p.rates
+
+  @inbounds fill_cur_rates(u, params, t, cur_rates, 1, rates...)
+  @inbounds cur_rates[1] = cur_rates[1]
```
Comment (isaacsas) [inline src/aggregators/direct.jl]: Is this necessary?

File: src/aggregators/direct.jl [hunk 3]
```diff
@@ -1,68 +1,94 @@
...
-  @fastmath normalizer = 1/sum_rate
-  @inbounds cur_rates[1] = normalizer*cur_rates[1]
-  @inbounds for i in 2:length(cur_rates) # normalize for choice, cumsum
-    cur_rates[i] = normalizer*cur_rates[i] + cur_rates[i-1]
-  end
-  sum_rate,randexp()/sum_rate
+############################# Required Functions #############################
+
+# creating the JumpAggregation structure
+@inline function aggregate(aggregator::Direct, u, p, t, end_time, constant_jumps, save_positions, rng)
```
Comment (isaacsas) [inline src/aggregators/direct.jl]: The inline here also seems unnecessary (it was present before but I don't see what good it can do here).

Comment (ChrisRackauckas) [reply src/aggregators/direct.jl]: Yeah it's probably unnecessary.

File: src/problem.jl (hunk 1 above)
Comment (isaacsas) [reply src/problem.jl]: It failed a different series of tests (newer I think) called PractRand. I'm not an expert on these tests, so don't know the significance. You can see the details on the blog post I linked back in the issue thread.

**Comment by isaacsas** [comment]
It looks good to me. This is the basic interface we discussed in the issue thread. I'll send through the mass action updates once this is in.

File: src/aggregators/direct.jl [hunk 4]
```diff
@@ -1,68 +1,93 @@
...
-    cur_rates[i] = normalizer*cur_rates[i] + cur_rates[i-1]
-  end
-  sum_rate,randexp()/sum_rate
+############################# Required Functions #############################
+
+# creating the JumpAggregation structure
+function aggregate(aggregator::Direct, u, p, t, end_time, constant_jumps, save_positions, rng)
+  rates = ((c.rate for c in constant_jumps)...)
+  affects! = ((c.affect! for c in constant_jumps)...)
+  cur_rates = Vector{Float64}(length(rates))
```
Comment (isaacsas) [inline src/aggregators/direct.jl]: Maybe instead of using Float64's here you should use `typeof(t)`?

Comment (alanderos91) [reply src/aggregators/direct.jl]: Sure. I'll hold off on pushing so I don't trigger so many builds...

File: src/problem.jl (hunk 1 above)
Comment (alanderos91) [reply src/problem.jl]: I don't know enough to comment on this but can change as needed.

File: src/aggregators/direct.jl (hunk 4 above)
Comment (isaacsas) [reply src/aggregators/direct.jl]: Yeah, sorry for the intermittent stream of comments there. I'm still getting used to looking at PRs on Github... That should be it from me -- I've gone through everything now.

Comment (ChrisRackauckas) [reply src/aggregators/direct.jl]: Don't worry about triggering builds. It'll queue.

Comment (ChrisRackauckas) [reply src/aggregators/direct.jl]: Note you can also add [ci-skip] to make it not queue a build for a commit.

**Comment by alanderos91** [comment]
Fantastic. Many thanks to both you for all the help! 🎉

**Comment by ChrisRackauckas** [comment]
Many thanks to you!

**Pull request merged**